module/builder/collection/builder.go

```diff
@@ -248,22 +207,26 @@ func (b *Builder) BuildOn(parentID flow.Identifier, setter func(*flow.Header) er
 		}
 
 		// disallow un-finalized reference blocks
-		if refChainFinalizedHeight < refHeader.Height {
+		if refHeader.Height > buildCtx.refChainFinalizedHeight {
+			continue
+		}
+		// disallow reference blocks above the final block of the epoch
+		if buildCtx.refEpochHasEnded && refHeader.Height > buildCtx.refEpochFinalHeight {
 			continue
 		}
 		// make sure the reference block is finalized and not orphaned
-		blockFinalizedAtReferenceHeight, err := b.mainHeaders.ByHeight(refHeader.Height)
+		blockIDFinalizedAtRefHeight, err := b.mainHeaders.BlockIDByHeight(refHeader.Height)
 		if err != nil {
 			return nil, fmt.Errorf("could not check that reference block (id=%x) is finalized: %w", tx.ReferenceBlockID, err)
 		}
-		if blockFinalizedAtReferenceHeight.ID() != tx.ReferenceBlockID {
+		if blockIDFinalizedAtRefHeight != tx.ReferenceBlockID {
 			// the transaction references an orphaned block - it will never be valid
 			b.transactions.Remove(tx.ID())
```

Comment:
would suggest to move line 235 further up to avoid repeated ID computation https://github.com/onflow/flow-go/blob/9413947ed37952ad13e3ca58919585f5b3361b94/module/builder/collection/builder.go#L235

Reply:
We won't actually repeat the ID computation, because these indented `tx.ID()` calls are all followed by `continue` statements. I've moved the main ID computation before them anyway to make this more clear.